src/java/org/apache/cassandra/service/accord/AccordSerializers.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.cassandra.db.marshal.CollectionType;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.db.marshal.MapType;
+import org.apache.cassandra.db.marshal.SetType;
+import org.apache.cassandra.db.partitions.PartitionUpdate;
+import org.apache.cassandra.db.rows.DeserializationHelper;
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputBuffer;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.net.MessagingService;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import static com.google.common.primitives.Ints.checkedCast;
+import static org.apache.cassandra.db.TypeSizes.sizeofUnsignedVInt;
+import static org.apache.cassandra.db.marshal.CollectionType.Kind.LIST;
+import static org.apache.cassandra.db.marshal.CollectionType.Kind.MAP;
+import static org.apache.cassandra.db.marshal.CollectionType.Kind.SET;
+
+public class AccordSerializers
+{
+    public static <T> ByteBuffer serialize(T item, IVersionedSerializer<T> serializer)
+    {
+        int version = MessagingService.current_version;
+        long size = serializer.serializedSize(item, version) + sizeofUnsignedVInt(version);
+        try (DataOutputBuffer out = new DataOutputBuffer((int) size))
+        {
+            out.writeUnsignedVInt(version);
+            serializer.serialize(item, out, version);
+            return out.buffer(false);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static <T> T deserialize(ByteBuffer bytes, IVersionedSerializer<T> serializer)
+    {
+        try (DataInputBuffer in = new DataInputBuffer(bytes, true))
+        {
+            int version = checkedCast(in.readUnsignedVInt());
+            return serializer.deserialize(in, version);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static Term.Terminal deserializeCqlCollectionAsTerm(ByteBuffer buffer, AbstractType<?> type)
+    {
+        CollectionType<?> collectionType = (CollectionType<?>) type;
+
+        if (collectionType.kind == SET)
+            return Sets.Value.fromSerialized(buffer, (SetType<?>) type, ProtocolVersion.CURRENT);
+        else if (collectionType.kind == LIST)
+            return Lists.Value.fromSerialized(buffer, (ListType<?>) type, ProtocolVersion.CURRENT);
+        else if (collectionType.kind == MAP)
+            return Maps.Value.fromSerialized(buffer, (MapType<?, ?>) type, ProtocolVersion.CURRENT);
+
```

Comment:
I should have just used V3 here, since that's what everything else does w/ collections. Given that's the case, it seems like we should just introduce a "default collection serialization version" or something somewhere, but I'd need to see how many more files this patch would touch...

Reply:
Passing the protocol version through `TxnReferenceOperation#apply()` isn't too hard. A little more work for `TxnCondition#applies()`...